### PR TITLE
add: initial simple conversion gb/sbol3 with sample files

### PR DESCRIPTION
**Pull request by mohitdmak**
### Changes:
* The conversion function takes in the GenBank input file and sbol3 file to write to, though not from the main function directly, it is hardcoded for now as we have it working only for a simple conversion having only 1 sequence in the GenBank file.
* I have included a separate file for unit tests pertaining to this direct conversion. Note that the conversion utility itself is also in a separate file at the moment in order to:
    - Not disturb the existing online conversion.
    - Better handle all test cases as we will probably make many helper functions, and they, along with the conversion function can be merged at the end into the `sbol_utilities.conversion.py` file.
### TODO:
* Make the conversion function available as a console script for the package - may be in a separate issue?

### CODE REVIEW

**Comment by tcmitchell** [comment]
@mohitdmak I'm guessing this is still a work in progress given the description of "TODO". Let us know when you're ready for review by adding @jakebeal and @tcmitchell as reviewers.

**Comment by mohitdmak** [comment]
@tcmitchell  @jakebeal Yes, it wasn't quite ready before, I have added the unit tests and have updated the description, but for some reason cannot add reviewers myself to the PR.

**Comment by tcmitchell** [comment]
@mohitdmak I have added reviewers (@jakebeal).

@mohitdmak Do you know why the unit test you added is failing? Do you need help with that?

**Comment by tcmitchell** [comment]
@mohitdmak I'm getting odd results with the unit tests via `pytest`. When I run `test/test_genbank_sbol3_direct.py` alone it passes:
```shell
$ pytest test/test_genbank_sbol3_direct.py             
======================================================== test session starts ========================================================
platform darwin -- Python 3.10.2, pytest-7.1.1, pluggy-1.0.0
rootdir: /Users/tmitchel/Projects/sd2/SBOL-utilities
collected 1 item                                                                                                                    

test/test_genbank_sbol3_direct.py .                                                                                           [100%]

========================================================= warnings summary ==========================================================
venv/lib/python3.10/site-packages/pyshacl/monkey/__init__.py:3
  /Users/tmitchel/Projects/sd2/SBOL-utilities/venv/lib/python3.10/site-packages/pyshacl/monkey/__init__.py:3: DeprecationWarning: The distutils package is deprecated and slated for removal in Python 3.12. Use setuptools or check PEP 632 for potential alternatives
    from distutils.version import LooseVersion

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
=================================================== 1 passed, 1 warning in 0.19s ====================================================
```

When I run `test/test_genbank_sbol3_direct.py` with other tests it fails:

```shell
pytest test/test_expand_combinatorial_derivations.py test/test_genbank_sbol3_direct.py
======================================================== test session starts ========================================================
platform darwin -- Python 3.10.2, pytest-7.1.1, pluggy-1.0.0
rootdir: /Users/tmitchel/Projects/sd2/SBOL-utilities
collected 4 items                                                                                                                   

test/test_expand_combinatorial_derivations.py ...                                                                             [ 75%]
test/test_genbank_sbol3_direct.py F                                                                                           [100%]

============================================================= FAILURES ==============================================================
_______________________________________________ TestGenBank2SBOL3.test_simple_file_1 ________________________________________________

self = <test_genbank_sbol3_direct.TestGenBank2SBOL3 testMethod=test_simple_file_1>

    def test_simple_file_1(self):
        """Test conversion of a simple genbank file with a single sequence"""
        TEST_OUTPUT_SBOL3 = SBOL3_FILE_1 + ".test"
        # Don't write to file for testing, we directly compare sbol documents
        test_output_sbol3 = convert_genbank_to_sbol3(GENBANK_FILE_1, TEST_OUTPUT_SBOL3, write=False)
        sbol3_file_1 = sbol3.Document()
        sbol3_file_1.read(location=SBOL3_FILE_1, file_format=sbol3.SORTED_NTRIPLES)
>       assert not doc_diff(test_output_sbol3, sbol3_file_1), \
            f'Converted SBOL3 file: {TEST_OUTPUT_SBOL3} not identical to expected file: {SBOL3_FILE_1}'
E       AssertionError: Converted SBOL3 file: /Users/tmitchel/Projects/sd2/SBOL-utilities/test/test_files/BBa_J23101_from_genbank_to_sbol3_direct.nt.test not identical to expected file: /Users/tmitchel/Projects/sd2/SBOL-utilities/test/test_files/BBa_J23101_from_genbank_to_sbol3_direct.nt
E       assert not 1
E        +  where 1 = doc_diff(<sbol3.document.Document object at 0x10a315d50>, <sbol3.document.Document object at 0x10a317ca0>)

test/test_genbank_sbol3_direct.py:16: AssertionError
------------------------------------------------------- Captured stdout call --------------------------------------------------------
Triples in Document 1, not in Document 2:
	http://sbolstandard.org/testfiles/BBa_J23101, http://sbols.org/v3#role, https://identifiers.org/SO:0000804
	http://sbolstandard.org/testfiles/BBa_J23101, http://sbols.org/v3#type, https://identifiers.org/SO:0000987
	http://sbolstandard.org/testfiles/BBa_J23101_sequence, http://sbols.org/v3#hasNamespace, http://sbolstandard.org/testfiles
	http://sbolstandard.org/testfiles/BBa_J23101, http://sbols.org/v3#description, constitutive promoter family member
	http://sbolstandard.org/testfiles/BBa_J23101_sequence, http://www.w3.org/1999/02/22-rdf-syntax-ns#type, http://sbols.org/v3#Sequence
	http://sbolstandard.org/testfiles/BBa_J23101_sequence, http://sbols.org/v3#elements, tttacagctagctcagtcctaggtattatgctagc
	http://sbolstandard.org/testfiles/BBa_J23101, http://sbols.org/v3#hasSequence, http://sbolstandard.org/testfiles/BBa_J23101_sequence
	http://sbolstandard.org/testfiles/BBa_J23101_sequence, http://sbols.org/v3#encoding, https://identifiers.org/edam:format_1207
	http://sbolstandard.org/testfiles/BBa_J23101_sequence, http://sbols.org/v3#displayId, BBa_J23101_sequence
	http://sbolstandard.org/testfiles/BBa_J23101, http://sbols.org/v3#type, https://identifiers.org/SBO:0000251
	http://sbolstandard.org/testfiles/BBa_J23101, http://sbols.org/v3#displayId, BBa_J23101
	http://sbolstandard.org/testfiles/BBa_J23101, http://www.w3.org/1999/02/22-rdf-syntax-ns#type, http://sbols.org/v3#Component
	http://sbolstandard.org/testfiles/BBa_J23101, http://sbols.org/v3#hasNamespace, http://sbolstandard.org/testfiles
Triples in Document 2, not in Document 1:
	https://testing.sbol3.genbank/BBa_J23101, http://sbols.org/v3#hasNamespace, https://testing.sbol3.genbank/
	https://testing.sbol3.genbank/BBa_J23101_sequence, http://www.w3.org/1999/02/22-rdf-syntax-ns#type, http://sbols.org/v3#Sequence
	https://testing.sbol3.genbank/BBa_J23101_sequence, http://sbols.org/v3#encoding, https://identifiers.org/edam:format_1207
	https://testing.sbol3.genbank/BBa_J23101, http://sbols.org/v3#type, https://identifiers.org/SBO:0000251
	https://testing.sbol3.genbank/BBa_J23101, http://www.w3.org/1999/02/22-rdf-syntax-ns#type, http://sbols.org/v3#Component
	https://testing.sbol3.genbank/BBa_J23101_sequence, http://sbols.org/v3#displayId, BBa_J23101_sequence
	https://testing.sbol3.genbank/BBa_J23101_sequence, http://sbols.org/v3#hasNamespace, https://testing.sbol3.genbank/
	https://testing.sbol3.genbank/BBa_J23101, http://sbols.org/v3#hasSequence, https://testing.sbol3.genbank/BBa_J23101_sequence
	https://testing.sbol3.genbank/BBa_J23101, http://sbols.org/v3#type, https://identifiers.org/SO:0000987
	https://testing.sbol3.genbank/BBa_J23101_sequence, http://sbols.org/v3#elements, tttacagctagctcagtcctaggtattatgctagc
	https://testing.sbol3.genbank/BBa_J23101, http://sbols.org/v3#displayId, BBa_J23101
	https://testing.sbol3.genbank/BBa_J23101, http://sbols.org/v3#description, constitutive promoter family member
	https://testing.sbol3.genbank/BBa_J23101, http://sbols.org/v3#role, https://identifiers.org/SO:0000804
========================================================= warnings summary ==========================================================
venv/lib/python3.10/site-packages/pyshacl/monkey/__init__.py:3
  /Users/tmitchel/Projects/sd2/SBOL-utilities/venv/lib/python3.10/site-packages/pyshacl/monkey/__init__.py:3: DeprecationWarning: The distutils package is deprecated and slated for removal in Python 3.12. Use setuptools or check PEP 632 for potential alternatives
    from distutils.version import LooseVersion

test/test_expand_combinatorial_derivations.py: 188 warnings
  /Users/tmitchel/Projects/sd2/SBOL-utilities/venv/lib/python3.10/site-packages/sbol3/toplevel.py:176: DeprecationWarning: Use sbol3.copy() instead
    warnings.warn('Use sbol3.copy() instead', DeprecationWarning)

test/test_expand_combinatorial_derivations.py: 792 warnings
  /Users/tmitchel/Projects/sd2/SBOL-utilities/venv/lib/python3.10/site-packages/sbol3/object.py:101: DeprecationWarning: Use sbol3.copy() instead
    warnings.warn('Use sbol3.copy() instead', DeprecationWarning)

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
====================================================== short test summary info ======================================================
FAILED test/test_genbank_sbol3_direct.py::TestGenBank2SBOL3::test_simple_file_1 - AssertionError: Converted SBOL3 file: /Users/tmi...
============================================ 1 failed, 3 passed, 981 warnings in 11.09s =============================================
```

I don't know what to make of this. Does this work in your environment?

File: sbol_utilities/sbol3_genbank_conversion.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
+import os
+import sbol3
+sbol3.set_namespace("https://testing.sbol3.genbank/")
```
Comment (tcmitchell) [inline sbol_utilities/sbol3_genbank_conversion.py:3]: This does not belong in the converter. A user should set this to the namespace they want. The converter should honor what the user has chosen.

File: sbol_utilities/sbol3_genbank_conversion.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+from Bio import GenBank
+GENBANK_PARSER = GenBank.RecordParser()
+
+# Conversion Constants (NOTE: most are placeholding and temporary for now)
+SAMPLE_GB_FILE_1 = "BBa_J23101.gb"          # Simplest file in genbank consisting only of a singular sequence
+SAMPLE_SBOL3_FILE_1 = "BBa_J23101_from_genbank_to_sbol3_direct.nt"
+COMP_TYPES = [sbol3.SBO_DNA]                # Temporarily assuming only dna components to be dealt with in genbank files
+COMP_ROLES = [sbol3.SO_ENGINEERED_REGION]   # Temporarilty assuming components to only have the engineered_region role
+GENBANK_FILE_1 = os.path.join(os.path.abspath(os.path.curdir), \
```
Comment (tcmitchell) [inline sbol_utilities/sbol3_genbank_conversion.py:12]: backslashes are not needed within parentheses. Please remove them here and elsewhere in this file.

File: test/test_genbank_sbol3_direct.py [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+class TestGenBank2SBOL3(unittest.TestCase):
+    def test_simple_file_1(self):
+        """Test conversion of a simple genbank file with a single sequence"""
+        sbol3.set_namespace("https://testing.sbol3.genbank/")
+        TEST_OUTPUT_SBOL3 = SBOL3_FILE_1 + ".test" 
+        # Don't write to file for testing, we directly compare sbol documents
+        test_output_sbol3 = convert_genbank_to_sbol3(GENBANK_FILE_1, TEST_OUTPUT_SBOL3, write=False)
+        sbol3_file_1 = sbol3.Document()
+        sbol3_file_1.read(location=SBOL3_FILE_1, file_format=sbol3.SORTED_NTRIPLES)
+        assert not doc_diff(test_output_sbol3, sbol3_file_1), \
```
Comment (tcmitchell) [inline test/test_genbank_sbol3_direct.py:16]: unnecessary backslash here as well

File: sbol_utilities/sbol3_genbank_conversion.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+                    roles=COMP_ROLES, \
+                    description=record.definition)
+    doc.add(comp)
+    # NOTE: Currently we use a single method of encoding (IUPAC)
+    seq = sbol3.Sequence(identity=record.accession[0] + "_sequence", \
+                    elements=record.sequence.lower(), \
+                    encoding=sbol3.IUPAC_DNA_ENCODING)
+    doc.add(seq)
+    comp.sequences = [ seq ]
+    if write: doc.write(fpath=sbol3_file, file_format=sbol3.SORTED_NTRIPLES)
```
Comment (tcmitchell) [inline sbol_utilities/sbol3_genbank_conversion.py:34]: Give `if write:` its own line and put the `doc.write` call on the next line.

File: sbol_utilities/sbol3_genbank_conversion.py [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+                    types=COMP_TYPES, \
+                    roles=COMP_ROLES, \
+                    description=record.definition)
+    doc.add(comp)
+    # NOTE: Currently we use a single method of encoding (IUPAC)
+    seq = sbol3.Sequence(identity=record.accession[0] + "_sequence", \
+                    elements=record.sequence.lower(), \
+                    encoding=sbol3.IUPAC_DNA_ENCODING)
+    doc.add(seq)
+    comp.sequences = [ seq ]
```
Comment (tcmitchell) [inline sbol_utilities/sbol3_genbank_conversion.py:33]:
```suggestion
    comp.sequences = [seq]
```

**Review by tcmitchell — CHANGES_REQUESTED** [review]
I have added a few comments to help clean up this pull request. In general I encourage you to use `pycodestyle` to help with Python style so that your code conforms to common python standards. The code does not have to precisely follow pycodestyle, and we should have good reasons to ignore pycodestyle warnings. Longer line length is one of those reasons.

File: sbol_utilities/sbol3_genbank_conversion.py (hunk 1 above)
Comment (mohitdmak) [reply sbol_utilities/sbol3_genbank_conversion.py]: Alright, I have provided namespace as a parameter to the function with a default.

File: test/test_genbank_sbol3_direct.py (hunk 3 above)
Comment (mohitdmak) [reply test/test_genbank_sbol3_direct.py:17]: For this case, actually it's outside parentheses, and the linter gives an error on removing it.

**Comment by mohitdmak** [comment]
> @mohitdmak I have created a branch on SBOL-utilities named `feature/sbol3-genbank-conversion` and I have retargeted this pull request onto that branch. This will be a long-running feature branch to hold your changes, and you should target pull requests to `feature/sbol3-genbank-conversion` instead of `develop` until we get a converter that can credibly be integrated with SBOL-utilities.

Noted, I have fixed the code style and namespaces options fix, also currently few other unit tests may be failing as some Http requests are getting timed out.

**Review by tcmitchell — APPROVED** [review]
Further stylistic cleanups can be done in future iterations. This looks good for now.

**Pull request merged**